### PR TITLE
fix(platform-browser): prevent clobbered elements from freezing the browser

### PR DESCRIPTION
**Pull request by vicb**
/cc @petebacondarwin @gkalpak @IgorMinar @mprobst 

Pete, George could one of you review the change ?

Thanks

### CODE REVIEW

**Comment by vicb** [comment]
@petebacondarwin @gkalpak tests are red on IE11, IE mobile 11 and Edge 14.14393.0, any idea ?

**Review by gkalpak — COMMENTED** [review]
LGTM
No idea about the failures. The only thing that is different with IE's `contains` is [this][1], but it shouldn't affect the test cases.

[1]: https://connect.microsoft.com/IE/feedback/details/780874/node-contains-is-incorrect

File: packages/platform-server/src/parse5_adapter.ts [hunk 1]
```diff
@@ -48,6 +48,15 @@ export class Parse5DomAdapter extends DomAdapter {
     setRootDomAdapter(new Parse5DomAdapter());
   }
 
+  contains(nodeA: any, nodeB: any): boolean {
+    let inner = nodeB;
+    while (inner) {
+      if (inner === nodeA) return true;
+      inner = inner.parent;
```
Comment (vicb) [inline packages/platform-server/src/parse5_adapter.ts]: yep, see `parentElement(el: any): Node { return el.parent; }`

**Comment by vicb** [comment]
@gkalpak Do you test angular.js on those failing browsers ? There was no CI link in the [ng1 PR](https://github.com/angular/angular.js/pull/15699) I'm not sure how you run the tests.

As said there is no timeout (hence no infinite loop) on IE/Edge where the tests are failing, I think it's safe to relax the tests. Any comment/objection ?

**Comment by petebacondarwin** [comment]
@vicb we do test against IE11 in AngularJS so I am not sure why the version here would fail when the version in AngularJS passes.

File: packages/platform-server/src/parse5_adapter.ts (hunk 1 above)
Comment (gkalpak) [inline packages/platform-server/src/parse5_adapter.ts]: I am still not convinced :)
I can't seem to be able to find a `parent` property neither in the HTML5 spec (which parse5 is supposed to adhere to) nor when running `parse5` locally. I may have missed something, though.

**Comment by gkalpak** [comment]
No idea why there is no link to Travis on the PR. You can find the latest tests run on Travis here: https://travis-ci.org/angular/angular.js/builds

From the affected browsers we only run tests on IE11.

**Review by IgorMinar — APPROVED** [review]
Please respond to @gkalpak's comment about parse5 before sending this for merging

File: packages/platform-server/src/parse5_adapter.ts (hunk 1 above)
Comment (vicb) [inline packages/platform-server/src/parse5_adapter.ts]: `parentElement(el: any): Node { return el.parent; }` is from the adapter code, if they were no `.parent` nothing would work.

But I hear you're still not convinced, so:
- we use the [`htmlparser2`](https://github.com/angular/angular/blob/master/packages/platform-server/src/parse5_adapter.ts#L44) tree adapter
- see the [`parent`](https://github.com/inikulin/parse5/blob/master/lib/tree_adapters/htmlparser2.js#L110) field for els in the source code and some [usage](https://github.com/inikulin/parse5/blob/master/lib/tree_adapters/htmlparser2.js#L273)

Happy to have convinced you :)

**Comment by vicb** [comment]
@chuckjaz ready. thanks.

File: packages/platform-server/src/parse5_adapter.ts (hunk 1 above)
Comment (gkalpak) [inline packages/platform-server/src/parse5_adapter.ts]: Convinced :smiley:

**Pull request merged**